### PR TITLE
graceful termination controller

### PR DESCRIPTION
**Pull request by roffe**
This PR adds a graceful termination controller by replacing the ipvs destination delete command. In case of restarts it's self-healing as NSC will detect destinations that shouldn't be on the service and queue them up for deletion after `ipvs-graceful-period` has passed or if there are 0 Active & Inactive connections when the sync loop is ran

Closes https://github.com/cloudnativelabs/kube-router/issues/572

### CODE REVIEW

File: pkg/controllers/graceful/graceful_termination_controller.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		if req.ipvsSvc.Address.String() == dst.ipvsSvc.Address.String() && req.ipvsSvc.Port == dst.ipvsSvc.Port && req.ipvsSvc.Protocol == dst.ipvsSvc.Protocol {
+			if req.ipvsDst.Address.String() == dst.ipvsDst.Address.String() && req.ipvsDst.Port == dst.ipvsDst.Port {
+				glog.V(2).Infof("Deletion request exists for svc: %v dst: %v", *req.ipvsSvc, *req.ipvsDst)
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		// Set the destination weight to 0
```
[the remark this answers is not in the document]
Comment (roffe) [reply pkg/controllers/graceful/graceful_termination_controller.go]: I set the weight to 0 so no new connections get sent to the backend destination, while old ones are not closed because a pod is being flagged as shutting down ( good example would be rolling upgrade a webservers with graceful shutdown enabled )

File: pkg/controllers/graceful/graceful_termination_controller.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	}
+	glog.V(1).Infof("Deleted conntrack entry for endpoint: %s:%d", dest.ipvsSvc.Address.String(), dest.ipvsSvc.Port)
+	return nil
+}
+
+// handleReq is the function that processes incoming messages on the job queue
+func (gh *TerminationController) handleReq(req gracefulRequest) error {
+	var found bool
+	for _, dst := range gh.jobQueue {
+		if req.ipvsSvc.Address.String() == dst.ipvsSvc.Address.String() && req.ipvsSvc.Port == dst.ipvsSvc.Port && req.ipvsSvc.Protocol == dst.ipvsSvc.Protocol {
```
[the remark this answers is not in the document]
Comment (roffe) [reply pkg/controllers/graceful/graceful_termination_controller.go]: Gotcha

[the remark this answers is not in the document]
Comment (roffe) [reply pkg/controllers/graceful/graceful_termination_controller.go]: Address has a Equal() that returns a bool that i can use

**Comment by roffe** [comment]
rebased

**Comment by murali-reddy** [comment]
@roffe quick comment. could you please keep `graceful_termination_controller.go` as part of `pkg/controllers/proxy/`. Its neither a cross-cutting concern nor significant enough to keep it separate IMO.

**Comment by roffe** [comment]
@murali-reddy done

**Comment by roffe** [comment]
rebased on latest master again

**Comment by murali-reddy** [comment]
@roffe Nice work!! couple of quick questions

- So what happens if kube-router pod is restarted when there are number of jobs in-flight?
- is there a race between actual pod deletion and kube-router waiting for connections to drain and remove the endpoint from the service. Is it possible that pod gets deleted by kubelet before `--ipvs-graceful-period` time?

**Comment by roffe** [comment]
@murali-reddy I see your point. i can see pod flapping causing trouble here where there will be a job to remove the destination after it has come back. I can solve that by tapping in to the add IPVS backend and have it remove any existing jobs for the destination in question if it gets re-added while being scheduled for graceful termination

backends can be deleted before `--ipvs-graceful-period` time, if the active & inactive conns is 0 when the backend is evaluated for removal in the controller. this currently occurs on a 10 second fixed rate.

On a sidenote, should i make a cli-opt to adjust this tick-rate for the controller?

**Comment by murali-reddy** [comment]
@roffe sorry you have to give me some more time to fully understand the what would be the graceful termination from Kuberentes perspective and for me to give meaningful feedback.

This is from https://kubernetes.io/docs/concepts/workloads/pods/pod/

```
By default, all deletes are graceful within 30 seconds. The kubectl delete command supports the --grace-period=<seconds> option which allows a user to override the default and specify their own value. The value 0 force deletes the pod. In kubectl version >= 1.5, you must specify an additional flag --force along with --grace-period=0 in order to perform force deletions.
```
I think we should be fully complaint with prescribed semantics (for e.g force delete)

**Comment by roffe** [comment]
@murali-reddy i've also discovered other obstacles, for some reason the traffic dies to the pod even though i've set up a scenario with propper livliness checks and the IPVS service gettings it's weight adjusted to 0. i suspect this is due to the routes controller withdrawing any kernel routes for the pod on the node killing the connection

**Comment by murali-reddy** [comment]
@roffe I tried to follow through the PR. Its a big change. Its hard to follow through the changes and rational. I am not sure it would have been possible to submit smaller incremental PR's. May be first submit the PR's for the issues that you surfaced which are independent of the graceful termination functionality?

IMO we should not have `--ipvs-graceful-period`. We can not generalise it for all workloads. We should assume default of 30 sec, unless its overwritten by `terminationGracePeriodSeconds` value in the pod specification.

Could you explain rational behind replacing `nsc.ln.ipvsAddService` with `nsc.gracefulController.AddService`?

**Comment by roffe** [comment]
We need to rewrite NSC from scratch i found several bugs in it relating to how it handles services which more or less make graceful handling impossible
